packages/vite/src/node/packages.ts

```diff
@@ -217,34 +215,43 @@ export function loadPackageData(pkgPath: string): PackageData {
   return pkg
 }
 
-export function watchPackageDataPlugin(config: ResolvedConfig): Plugin {
+export function watchPackageDataPlugin(packageCache: PackageCache): Plugin {
+  // a list of files to watch before the plugin is ready
   const watchQueue = new Set<string>()
-  let watchFile = (id: string) => {
+  const watchedDirs = new Set<string>()
+
+  const watchFileStub = (id: string) => {
     watchQueue.add(id)
   }
+  let watchFile = watchFileStub
 
-  const { packageCache } = config
   const setPackageData = packageCache.set.bind(packageCache)
   packageCache.set = (id, pkg) => {
-    if (id.endsWith('.json')) {
-      watchFile(id)
+    if (!isInNodeModules(pkg.dir) && !watchedDirs.has(pkg.dir)) {
+      watchedDirs.add(pkg.dir)
+      watchFile(path.join(pkg.dir, 'package.json'))
```

Comment:
Make sure to only watch `package.json` in `node_modules`, also uses a `watchedDirs` set to cache as this `.set` call can be blasted with many same paths. In dev, `watchFile` would invoke a `fs.existSync` call otherwise.